File: clients/go/corsaircloud.go
// Package corsaircloud is a thin HTTP client for a hosted Corsair Cloud
// project. The plugin set lives on the VM, so calls are dynamic.
package corsaircloud

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// Client talks to a Corsair Cloud project. The base URL is derived from the
// API key (https://api.corsair.cloud/<slug>/api/corsair), so the key is the
// only value you pass.
type Client struct {
	apiKey  string
	baseURL string
	http    *http.Client
	initErr error
}

// Option configures a Client.
type Option func(*Client)

// WithHTTPClient overrides the default http.Client.
func WithHTTPClient(h *http.Client) Option {
	return func(c *Client) { c.http = h }
}

// WithURL overrides the base URL derived from the key (dev/testing only).
func WithURL(u string) Option {
	return func(c *Client) { c.baseURL = u }
}

// urlFromKey derives the runtime URL from a ck_cloud_<slug>.<secret> key: the
// slug is the segment after the prefix up to the first '.' (the base64url
// secret never contains '.').
func urlFromKey(apiKey string) string {
	const prefix = "ck_cloud_"
	if !strings.HasPrefix(apiKey, prefix) {
		return ""
	}
	rest := apiKey[len(prefix):]
	i := strings.IndexByte(rest, '.')
	if i <= 0 {
		return ""
	}
	slug := rest[:i]
	for _, r := range slug {
		if !((r >= 'a' && r <= 'z') || (r >= '0' && r <= '9')) {
			return ""
		}
	}
	return "https://api.corsair.cloud/" + slug + "/api/corsair"
}

var loopbackHosts = map[string]bool{"localhost": true, "127.0.0.1": true, "::1": true}

// The API key is sent as a bearer token, so http:// would leak it in
// cleartext — allowed only for loopback, matching the other language clients.
func assertSecureBaseURL(baseURL string) error {
	u, err := url.Parse(baseURL)
	if err != nil {
		return fmt.Errorf("corsaircloud: invalid base URL %q: %w", baseURL, err)
	}
	if u.Scheme == "https" {
		return nil
	}
	if u.Scheme == "http" && loopbackHosts[u.Hostname()] {
		return nil
	}
	return fmt.Errorf("corsaircloud: base URL must use https:// (got %q) — http:// is only allowed for localhost/127.0.0.1", baseURL)
}

// New creates a Client for the given API key; the URL is derived from the key.
// Pass WithURL only for dev/testing. A URL that isn't https:// (loopback
// excepted), or a key no URL can be derived from, surfaces as an error from the
// first call made with this client.
func New(apiKey string, opts ...Option) *Client {
	c := &Client{apiKey: apiKey, http: &http.Client{Timeout: 30 * time.Second}}
	for _, opt := range opts {
		opt(c)
	}
	if c.baseURL == "" {
		c.baseURL = urlFromKey(apiKey)
	}
	if c.baseURL == "" {
		c.initErr = fmt.Errorf("corsaircloud: could not resolve a URL from apiKey — pass a ck_cloud_<slug>.<secret> key, or corsaircloud.WithURL(...)")
	} else {
		c.baseURL = strings.TrimRight(c.baseURL, "/")
		c.initErr = assertSecureBaseURL(c.baseURL)
	}
	return c
}

// Tenant scopes calls to a tenant for plugin op invocation.
func (c *Client) Tenant(id string) *TenantClient {
	return &TenantClient{client: c, tenantID: id}
}

// ConnectionStatus returns each plugin's connection state for a tenant.
func (c *Client) ConnectionStatus(ctx context.Context, tenantID string) (map[string]string, error) {
	data, err := c.send(ctx, http.MethodGet, []string{"connection-status"}, url.Values{"tenantId": {tenantID}}, nil)
	if err != nil {
		return nil, err
	}
	var out map[string]string
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// CreateConnectLink starts an OAuth connect flow for a plugin/tenant pair.
// redirectURI is optional — pass "" to omit it.
func (c *Client) CreateConnectLink(ctx context.Context, plugin, tenantID, redirectURI string) (ConnectLink, error) {
	body := map[string]string{"plugin": plugin, "tenantId": tenantID}
	if redirectURI != "" {
		body["redirectUri"] = redirectURI
	}
	data, err := c.send(ctx, http.MethodPost, []string{"connect", "links"}, nil, body)
	if err != nil {
		return ConnectLink{}, err
	}
	var out ConnectLink
	if err := json.Unmarshal(data, &out); err != nil {
		return ConnectLink{}, err
	}
	return out, nil
}

// Disconnect removes a plugin's credentials for a tenant.
func (c *Client) Disconnect(ctx context.Context, plugin, tenantID string) error {
	body := map[string]string{"plugin": plugin, "tenantId": tenantID}
	_, err := c.send(ctx, http.MethodPost, []string{"disconnect"}, nil, body)
	return err
}

// Tenants lists the project's tenants.
func (c *Client) Tenants(ctx context.Context) ([]Tenant, error) {
	data, err := c.send(ctx, http.MethodGet, []string{"tenants"}, nil, nil)
	if err != nil {
		return nil, err
	}
	var out []Tenant
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// CreateTenant creates a tenant with the given id.
func (c *Client) CreateTenant(ctx context.Context, id string) (Tenant, error) {
	data, err := c.send(ctx, http.MethodPost, []string{"tenants"}, nil, map[string]string{"id": id})
	if err != nil {
		return Tenant{}, err
	}
	var out Tenant
	if err := json.Unmarshal(data, &out); err != nil {
		return Tenant{}, err
	}
	return out, nil
}

// TenantClient invokes plugin ops scoped to one tenant.
type TenantClient struct {
	client   *Client
	tenantID string
}

// Call invokes a plugin op and returns the response's data field.
func (t *TenantClient) Call(ctx context.Context, plugin, op string, args any) (json.RawMessage, error) {
	if args == nil {
		// Send {"args":{}} rather than {"args":null}; the contract types args as
		// an object, matching the Python/Swift clients' empty-object default.
		args = map[string]any{}
	}
	raw, err := t.client.send(ctx, http.MethodPost, []string{t.tenantID, plugin, "call", op}, nil, map[string]any{"args": args})
	if err != nil {
		return nil, err
	}
	var env envelope
	if err := json.Unmarshal(raw, &env); err != nil {
		return nil, err
	}
	return env.Data, nil
}

// ConnectLink is a project's response to a connect-link request.
type ConnectLink struct {
	ConnectURL string `json:"connectUrl"`
	ExpiresAt  string `json:"expiresAt"`
	TenantID   string `json:"tenantId"`
}

// Tenant is a project tenant.
type Tenant struct {
	ID               string   `json:"id"`
	ConnectedPlugins []string `json:"connectedPlugins"`
}

// CorsairError is the runtime's error envelope plus the HTTP status.
// Code is the machine code (not_connected, unknown_plugin, provider_error, ...).
type CorsairError struct {
	Status         int
	Code           string
	Message        string
	Reason         string
	ProviderStatus int
}

func (e *CorsairError) Error() string {
	if e.Message != "" {
		return fmt.Sprintf("corsair: %s: %s", e.Code, e.Message)
	}
	return fmt.Sprintf("corsair: %s (status %d)", e.Code, e.Status)
}

type envelope struct {
	Data json.RawMessage `json:"data"`
}

type errorBody struct {
	Error          string `json:"error"`
	Message        string `json:"message"`
	Reason         string `json:"reason"`
	ProviderStatus int    `json:"providerStatus"`
}

func (c *Client) send(ctx context.Context, method string, path []string, query url.Values, body any) (json.RawMessage, error) {
	if c.initErr != nil {
		return nil, c.initErr
	}
	escaped := make([]string, len(path))
	for i, p := range path {
		escaped[i] = url.PathEscape(p)
	}
	u := c.baseURL + "/" + strings.Join(escaped, "/")
	if len(query) > 0 {
		u += "?" + query.Encode()
	}

	var reqBody *bytes.Reader
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		reqBody = bytes.NewReader(b)
	} else {
		reqBody = bytes.NewReader(nil)
	}

	req, err := http.NewRequestWithContext(ctx, method, u, reqBody)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+c.apiKey)
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := c.http.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	respBody := new(bytes.Buffer)
	if _, err := respBody.ReadFrom(resp.Body); err != nil {
		return nil, err
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		var eb errorBody
		if err := json.Unmarshal(respBody.Bytes(), &eb); err == nil && eb.Error != "" {
			return nil, &CorsairError{
				Status: resp.StatusCode, Code: eb.Error, Message: eb.Message,
				Reason: eb.Reason, ProviderStatus: eb.ProviderStatus,
			}
		}
		return nil, &CorsairError{Status: resp.StatusCode, Code: "http_error"}
	}

	return respBody.Bytes(), nil
}

File: clients/swift/Sources/CorsairCloud/CorsairCloud.swift
import Foundation

#if canImport(FoundationNetworking)
	import FoundationNetworking
#endif

/// A JSON value with no external dependency — carries op args in and results out.
public enum JSONValue: Codable, Sendable, Equatable {
	case string(String)
	case number(Double)
	case bool(Bool)
	case object([String: JSONValue])
	case array([JSONValue])
	case null

	public init(from decoder: Decoder) throws {
		let c = try decoder.singleValueContainer()
		if c.decodeNil() { self = .null; return }
		if let b = try? c.decode(Bool.self) { self = .bool(b); return }
		if let n = try? c.decode(Double.self) { self = .number(n); return }
		if let s = try? c.decode(String.self) { self = .string(s); return }
		if let o = try? c.decode([String: JSONValue].self) { self = .object(o); return }
		if let a = try? c.decode([JSONValue].self) { self = .array(a); return }
		throw DecodingError.dataCorruptedError(in: c, debugDescription: "unrecognized JSON value")
	}

	public func encode(to encoder: Encoder) throws {
		var c = encoder.singleValueContainer()
		switch self {
		case let .string(v): try c.encode(v)
		case let .number(v): try c.encode(v)
		case let .bool(v): try c.encode(v)
		case let .object(v): try c.encode(v)
		case let .array(v): try c.encode(v)
		case .null: try c.encodeNil()
		}
	}

	/// Re-decode this value into a concrete type (the typed `call` convenience).
	public func decoded<T: Decodable>(as type: T.Type) throws -> T {
		try JSONDecoder().decode(T.self, from: JSONEncoder().encode(self))
	}
}

extension JSONValue: ExpressibleByStringLiteral, ExpressibleByBooleanLiteral,
	ExpressibleByIntegerLiteral, ExpressibleByFloatLiteral
{
	public init(stringLiteral value: String) { self = .string(value) }
	public init(booleanLiteral value: Bool) { self = .bool(value) }
	public init(integerLiteral value: Int) { self = .number(Double(value)) }
	public init(floatLiteral value: Double) { self = .number(value) }
}

/// The runtime's error envelope (`{ error, message, reason?, providerStatus? }`)
/// plus the HTTP status. `code` is the machine code (`not_connected`, …).
public struct CorsairError: Error, Equatable, Sendable {
	public let status: Int
	public let code: String
	public let message: String?
	public let reason: String?
	public let providerStatus: Int?
}

/// Thrown at call time when `baseURL` isn't https:// (loopback excepted) — the
/// API key is sent as a bearer token, so http:// would leak it in cleartext.
public struct InsecureBaseURLError: Error, Equatable, Sendable {
	public let url: String
}

private let loopbackHosts: Set<String> = ["localhost", "127.0.0.1", "::1"]

private func assertSecureBaseURL(_ url: URL) throws {
	if url.scheme == "https" { return }
	if url.scheme == "http", let host = url.host, loopbackHosts.contains(host) { return }
	throw InsecureBaseURLError(url: url.absoluteString)
}

/// Thrown at call time when no URL was given and none could be derived from the
/// key — pass a `ck_cloud_<slug>.<secret>` key, or set `url` explicitly.
public struct UnresolvedURLError: Error, Equatable, Sendable {}

/// Thrown by a tenant call when the tenant id is empty — it would build a
/// request path with a missing segment and misroute (matches the TS client).
public struct EmptyTenantIdError: Error, Equatable, Sendable {}

private let cloudSlugChars = Set("abcdefghijklmnopqrstuvwxyz0123456789")

/// Derives the runtime URL from a ck_cloud_<slug>.<secret> key: the slug is the
/// segment after the prefix up to the first '.' (the base64url secret never
/// contains '.'), so the key is the only value you pass.
private func urlFromKey(_ apiKey: String) -> URL? {
	let prefix = "ck_cloud_"
	guard apiKey.hasPrefix(prefix) else { return nil }
	let rest = apiKey.dropFirst(prefix.count)
	guard let sep = rest.firstIndex(of: ".") else { return nil }
	let slug = rest[..<sep]
	guard !slug.isEmpty, slug.allSatisfy({ cloudSlugChars.contains($0) }) else { return nil }
	return URL(string: "https://api.corsair.cloud/\(slug)/api/corsair")
}

private func encodedPathSegment(_ segment: String) -> String {
	segment.addingPercentEncoding(withAllowedCharacters: .urlPathSegmentAllowed) ?? segment
}

extension CharacterSet {
	// Also escape "%" so a literal segment like "a%2Fb" encodes to "a%252Fb"
	// rather than staying "a%2Fb" and decoding to "a/b" on the server — matching
	// Go's url.PathEscape and Python's quote(safe="").
	fileprivate static let urlPathSegmentAllowed = CharacterSet.urlPathAllowed.subtracting(CharacterSet(charactersIn: "/%"))
}

public struct ConnectLink: Codable, Sendable {
	public let connectUrl: String?
	public let expiresAt: String?
	public let tenantId: String?
}

public struct Tenant: Codable, Sendable {
	public let id: String
	public let connectedPlugins: [String]?
}

/// Client for a hosted Corsair Cloud project. Mirrors `corsairCloud`: a
/// dynamic HTTP client — the plugin set lives on the VM.
public struct CorsairCloud: Sendable {
	let apiKey: String
	let baseURL: URL?
	let session: URLSession

	/// The URL is derived from the API key; pass `url` only for dev/testing.
	public init(apiKey: String, url: URL? = nil, session: URLSession = .shared) {
		self.apiKey = apiKey
		let resolved = url ?? urlFromKey(apiKey)
		if let resolved {
			// Strip a trailing slash so a console-copied base URL doesn't produce
			// a "//" in the request path, matching the Python/Go/TS clients.
			let s = resolved.absoluteString
			baseURL = URL(string: s.hasSuffix("/") ? String(s.dropLast()) : s) ?? resolved
		} else {
			baseURL = nil
		}
		self.session = session
	}

	public func tenant(_ id: String) -> TenantClient {
		TenantClient(client: self, tenantId: id)
	}

	public var manage: Manage { Manage(client: self) }

	func send(
		_ method: String,
		path: [String],
		query: [String: String] = [:],
		body: JSONValue? = nil
	) async throws -> Data {
		guard let baseURL else { throw UnresolvedURLError() }
		try assertSecureBaseURL(baseURL)
		var comps = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)!
		comps.percentEncodedPath += "/" + path.map(encodedPathSegment).joined(separator: "/")
		if !query.isEmpty {
			comps.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
		}
		var req = URLRequest(url: comps.url!)
		req.httpMethod = method
		req.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
		if let body {
			req.setValue("application/json", forHTTPHeaderField: "Content-Type")
			req.httpBody = try JSONEncoder().encode(body)
		}
		let (data, response) = try await session.data(for: req)
		let status = (response as? HTTPURLResponse)?.statusCode ?? 0
		guard (200 ..< 300).contains(status) else {
			if let e = try? JSONDecoder().decode(ErrorBody.self, from: data) {
				throw CorsairError(
					status: status, code: e.error, message: e.message,
					reason: e.reason, providerStatus: e.providerStatus)
			}
			throw CorsairError(
				status: status, code: "http_error", message: nil, reason: nil, providerStatus: nil)
		}
		return data
	}

	private struct ErrorBody: Decodable {
		let error: String
		let message: String?
		let reason: String?
		let providerStatus: Int?
	}

	struct Envelope<T: Decodable>: Decodable { let data: T }
}

public struct TenantClient: Sendable {
	let client: CorsairCloud
	let tenantId: String

	private func callRaw(
		_ plugin: String, _ op: String, _ args: [String: JSONValue]
	) async throws -> Data {
		guard !tenantId.isEmpty else { throw EmptyTenantIdError() }
		return try await client.send(
			"POST", path: [tenantId, plugin, "call", op], body: .object(["args": .object(args)]))
	}

	/// Invoke a plugin op, returning the raw result (`.data`) as a `JSONValue`.
	public func call(
		_ plugin: String, _ op: String, args: [String: JSONValue] = [:]
	) async throws -> JSONValue {
		let data = try await callRaw(plugin, op, args)
		return try JSONDecoder().decode(CorsairCloud.Envelope<JSONValue>.self, from: data).data
	}

	/// Invoke a plugin op, decoding `.data` into a concrete type.
	public func call<T: Decodable>(
		_ plugin: String, _ op: String, args: [String: JSONValue] = [:], as _: T.Type
	) async throws -> T {
		let data = try await callRaw(plugin, op, args)
		return try JSONDecoder().decode(CorsairCloud.Envelope<T>.self, from: data).data
	}
}

public struct Manage: Sendable {
	let client: CorsairCloud

	public func connectionStatus(tenantId: String) async throws -> [String: String] {
		let data = try await client.send(
			"GET", path: ["connection-status"], query: ["tenantId": tenantId])
		return try JSONDecoder().decode([String: String].self, from: data)
	}

	public func createConnectLink(
		plugin: String, tenantId: String, redirectUri: String? = nil
	) async throws -> ConnectLink {
		var body: [String: JSONValue] = ["plugin": .string(plugin), "tenantId": .string(tenantId)]
		if let redirectUri { body["redirectUri"] = .string(redirectUri) }
		let data = try await client.send("POST", path: ["connect", "links"], body: .object(body))
		return try JSONDecoder().decode(ConnectLink.self, from: data)
	}

	public func disconnect(plugin: String, tenantId: String) async throws {
		_ = try await client.send(
			"POST", path: ["disconnect"],
			body: .object(["plugin": .string(plugin), "tenantId": .string(tenantId)]))
	}

	public func tenants() async throws -> [Tenant] {
		try JSONDecoder().decode([Tenant].self, from: await client.send("GET", path: ["tenants"]))
	}

	public func createTenant(id: String) async throws -> Tenant {
		let data = try await client.send(
			"POST", path: ["tenants"], body: .object(["id": .string(id)]))
		return try JSONDecoder().decode(Tenant.self, from: data)
	}

	/// Fetch a permission record by id. The shape varies by grant, so it comes
	/// back as a `JSONValue` for the caller to read or decode.
	public func permission(id: String) async throws -> JSONValue {
		let data = try await client.send("GET", path: ["permissions", id])
		return try JSONDecoder().decode(JSONValue.self, from: data)
	}
}
